base/iterators.jl

```diff
@@ -988,7 +988,13 @@ cycle(xs, n::Integer) = flatten(repeated(xs, n))
 
 eltype(::Type{Cycle{I}}) where {I} = eltype(I)
 IteratorEltype(::Type{Cycle{I}}) where {I} = IteratorEltype(I)
-IteratorSize(::Type{Cycle{I}}) where {I} = IsInfinite() # XXX: this is false if iterator ever becomes empty
+function IteratorSize(::Type{Cycle{I}}) where I
+    # TODO: find a better way of communicating the size of a cycle
+    # IsInfinite() would be false if iterator ever becomes empty
+    IteratorSize(I) === IsInfinite() ? IsInfinite() : SizeUnknown()
```

Comment:
A refinement, to improve precision:

```suggestion
    c = IteratorSize(I)
    ((c === IsInfinite()) || (c === HasShape{0}())) ? IsInfinite() : SizeUnknown()
```

The idea is that there are two cases when we know the number of elements is not zero (and thus infinite):
* when `IteratorSize(I) === IsInfinite()`, already accounted for, just listing it for completeness)
* when `IteratorSize(I) === HasShape{0}()`, because a zero-dimensional shape implies a unit length